ocaml/xapi/xapi.ml

```diff
@@ -844,6 +844,29 @@ let init_tls_verification () =
       info "TLS verification is enabled: %s" file ;
       Stunnel_client.set_verify_by_default true
 
+let am_i_missing_certs_loop ~__context () =
+  let open Helpers in
+  let module Client = Client.Client in
+  let delay = 10.0 in
+  let inner () =
+    try
+      if
+        Db.Pool.get_tls_verification_enabled ~__context
+          ~self:(get_pool ~__context)
+        && Cert_distrib.am_i_missing_certs ~__context
+      then (
+        D.debug "am_i_missing_certs_loop: i am missing certs!" ;
+        let host = get_localhost ~__context in
+        call_api_functions ~__context @@ fun rpc session_id ->
+        Client.Host.copy_primary_host_certs ~rpc ~session_id ~host
+      )
+    with e ->
+      D.error "am_i_missing_certs_loop: exception (ignoring and continuing): %s"
+        (Printexc.to_string e)
+  in
+  let rec loop () = inner () ; Thread.delay delay ; (loop [@tailcall]) () in
```

Comment:
Why does there have to be a delay here, and if there has to be can you make it an exponential backoff that starts with a small delay (0.5s and ends up 10s at max?)
There is a Repeat_with_uniform_backoff in helpers.ml, could use that as a start and maybe add an exponential backoff policy later (I think there are exponential backoffs elsewhere in xapi already)

Reply:
I'm not sure I understand what advantage an expontential backoff would give us here? The idea of this loop is that if a slave's certificates become out of sync with what's the db, then it will ask the master for an update. In theory this we should only have to do this very rarely (it is an unfortunate solution for the cases of parallel pool join, or if certificate sync went wrong for some reason)

> Why does there have to be a delay here

i put the delay in so we are not querying the database and checking the file system in a tight loop. 10s was chosen arbitrarily